File: src/backend/compras/src/model/compra_fornecedor_cotacao_model.py
from src import db
from src.model.view_pessoa_fornecedor_model import ViewPessoaFornecedorModel


class CompraFornecedorCotacaoModel(db.Model):
    __tablename__ = 'compra_fornecedor_cotacao'

    id = db.Column(db.Integer, primary_key=True)
    codigo = db.Column(db.String(32))
    prazo_entrega = db.Column(db.String(50))
    venda_condicoes_pagamento = db.Column(db.String(50))
    valor_subtotal = db.Column(db.Float)
    taxa_desconto = db.Column(db.Float)
    valor_desconto = db.Column(db.Float)
    valor_total = db.Column(db.Float)
    id_fornecedor = db.Column(db.Integer, db.ForeignKey('view_pessoa_fornecedor.id'))
    id_compra_cotacao = db.Column(db.Integer, db.ForeignKey('compra_cotacao.id'))

    view_pessoa_fornecedor_model = db.relationship('ViewPessoaFornecedorModel', foreign_keys=[id_fornecedor])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_compra_cotacao = data.get('idCompraCotacao')
        self.id_fornecedor = data.get('idFornecedor')
        self.codigo = data.get('codigo')
        self.prazo_entrega = data.get('prazoEntrega')
        self.venda_condicoes_pagamento = data.get('vendaCondicoesPagamento')
        self.valor_subtotal = data.get('valorSubtotal')
        self.taxa_desconto = data.get('taxaDesconto')
        self.valor_desconto = data.get('valorDesconto')
        self.valor_total = data.get('valorTotal')

    def serialize(self):
        return {
            'id': self.id,
            'idCompraCotacao': self.id_compra_cotacao,
            'idFornecedor': self.id_fornecedor,
            'codigo': self.codigo,
            'prazoEntrega': self.prazo_entrega,
            'vendaCondicoesPagamento': self.venda_condicoes_pagamento,
            'valorSubtotal': self.valor_subtotal,
            'taxaDesconto': self.taxa_desconto,
            'valorDesconto': self.valor_desconto,
            'valorTotal': self.valor_total,
            'viewPessoaFornecedorModel': self.view_pessoa_fornecedor_model.serialize() if self.view_pessoa_fornecedor_model else None,
        }

File: src/frontend/compras/lib/app/data/model/compra_fornecedor_cotacao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:compras/app/data/model/model_imports.dart';

class CompraFornecedorCotacaoModel {
	int? id;
	int? idCompraCotacao;
	int? idFornecedor;
	String? codigo;
	String? prazoEntrega;
	String? vendaCondicoesPagamento;
	double? valorSubtotal;
	double? taxaDesconto;
	double? valorDesconto;
	double? valorTotal;
	ViewPessoaFornecedorModel? viewPessoaFornecedorModel;

	CompraFornecedorCotacaoModel({
		this.id,
		this.idCompraCotacao,
		this.idFornecedor,
		this.codigo,
		this.prazoEntrega,
		this.vendaCondicoesPagamento,
		this.valorSubtotal,
		this.taxaDesconto,
		this.valorDesconto,
		this.valorTotal,
		this.viewPessoaFornecedorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo',
		'prazo_entrega',
		'venda_condicoes_pagamento',
		'valor_subtotal',
		'taxa_desconto',
		'valor_desconto',
		'valor_total',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo',
		'Prazo Entrega',
		'Venda Condicoes Pagamento',
		'Valor Subtotal',
		'Taxa Desconto',
		'Valor Desconto',
		'Valor Total',
	];

	CompraFornecedorCotacaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCompraCotacao = jsonData['idCompraCotacao'];
		idFornecedor = jsonData['idFornecedor'];
		codigo = jsonData['codigo'];
		prazoEntrega = jsonData['prazoEntrega'];
		vendaCondicoesPagamento = jsonData['vendaCondicoesPagamento'];
		valorSubtotal = jsonData['valorSubtotal']?.toDouble();
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		valorDesconto = jsonData['valorDesconto']?.toDouble();
		valorTotal = jsonData['valorTotal']?.toDouble();
		viewPessoaFornecedorModel = jsonData['viewPessoaFornecedorModel'] == null ? ViewPessoaFornecedorModel() : ViewPessoaFornecedorModel.fromJson(jsonData['viewPessoaFornecedorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCompraCotacao'] = idCompraCotacao != 0 ? idCompraCotacao : null;
		jsonData['idFornecedor'] = idFornecedor != 0 ? idFornecedor : null;
		jsonData['codigo'] = codigo;
		jsonData['prazoEntrega'] = prazoEntrega;
		jsonData['vendaCondicoesPagamento'] = vendaCondicoesPagamento;
		jsonData['valorSubtotal'] = valorSubtotal;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['valorDesconto'] = valorDesconto;
		jsonData['valorTotal'] = valorTotal;
		jsonData['viewPessoaFornecedorModel'] = viewPessoaFornecedorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCompraCotacao = plutoRow.cells['idCompraCotacao']?.value;
		idFornecedor = plutoRow.cells['idFornecedor']?.value;
		codigo = plutoRow.cells['codigo']?.value;
		prazoEntrega = plutoRow.cells['prazoEntrega']?.value;
		vendaCondicoesPagamento = plutoRow.cells['vendaCondicoesPagamento']?.value;
		valorSubtotal = plutoRow.cells['valorSubtotal']?.value?.toDouble();
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		valorDesconto = plutoRow.cells['valorDesconto']?.value?.toDouble();
		valorTotal = plutoRow.cells['valorTotal']?.value?.toDouble();
		viewPessoaFornecedorModel = ViewPessoaFornecedorModel();
		viewPessoaFornecedorModel?.nome = plutoRow.cells['viewPessoaFornecedorModel']?.value;
	}	

	CompraFornecedorCotacaoModel clone() {
		return CompraFornecedorCotacaoModel(
			id: id,
			idCompraCotacao: idCompraCotacao,
			idFornecedor: idFornecedor,
			codigo: codigo,
			prazoEntrega: prazoEntrega,
			vendaCondicoesPagamento: vendaCondicoesPagamento,
			valorSubtotal: valorSubtotal,
			taxaDesconto: taxaDesconto,
			valorDesconto: valorDesconto,
			valorTotal: valorTotal,
		);			
	}

	
}